### PR TITLE
New User Onboarding Flow UI

### PR DESCRIPTION
**Pull request by selfcontained**
## Description

This adds a new onboarding flow for new users (currently we consider new users anyone that hasn't provided a preferred IDE). For those users, we'll show the following three step flow when they sign in.

| Step One      | Step Two | Step Three
| ----------- | ----------- | ----------- |
| image | image | image

* This is behind the `newSignupFlow` feature flag that is enabled in non-production.
* While we're collecting a distinct First and Last name in the flow, we're still storing these as a unified `fullName` on the user. Collecting as 2 fields is an attempt to help signal we're asking for real names and not aliases/usernames.
* Email is still stored on `user.additionalDetails.profile.emailAddress`
* Answers to step 3 are stored on new fields under `user.additionalDetails.profile`
* At the completion of the onboarding flow we're storing a timestamp on `user.additionalDetails.profile.onboardedTimestamp` to help us identify which user's have completed this flow in the future. We plan to introduce this flow, or a variation of it to existing users, and this will help us distinguish those that have gone through it.

## Related Issue(s)

Fixes #

## How to test

* Visit the preview environment: https://bmh-onboarding-form.preview.gitpod-dev.com/
* The first time you visit, you should be sent through the flow.
* You can verify your name and email changes are applied after completing the flow by visiting the User Settings.

**Repeat the flow**
You can also initiate the onboarding flow manually by appending the query param `?onboarding=force` to a url. For example: 

https://bmh-onboarding-form.preview.gitpod-dev.com/workspaces?onboarding=force 

... should kick you into the flow even if you've completed it. This was mainly added to aid in testing, but could be helpful in the future to send a user into the flow manually.

**With Starting Workspace links:**

You can also test this with a workspace context link and include the `?onboarding=force` query param:

https://bmh-onboarding-form.preview.gitpod-dev.com/?onboarding=force#https://github.com/gitpod-io/gitpod

## Release Notes

```release-note
We've introduced a new onboarding flow to help user's customize Gitpod and get to know their needs better.
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-slow-database
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`
- [x] /werft analytics=segment

### CODE REVIEW

**Comment by roboquat** [comment]
@selfcontained: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: components/dashboard/src/components/ThemeSelector.tsx
Comment (selfcontained) [inline components/dashboard/src/components/ThemeSelector.tsx:1]: Extracted this from `user-settings/Preferences.tsx` to re-use it.

File: components/dashboard/tsconfig.json
Comment (selfcontained) [inline components/dashboard/tsconfig.json:16]: Disabling for compilation to improve dx a bit. Once we have our linter enabled for PRs it will catch any unused vars rather than it be a compilation error.

File: components/dashboard/src/data/current-user/update-mutation.ts
Comment (Siddhant-K-code) [inline components/dashboard/src/data/current-user/update-mutation.ts:19]: **Suggestion (non-blocking):** How about we split it out in chunks, it can be also useful in the future to reuse the similar functionality any other place

```suggestion
export const useUpdateCurrentUserMutation = () => {
    return useMutation({
        mutationFn: updateUser,
    });
};

const updateUser = async (partialUser: UpdateCurrentUserArgs) => {
    return await getGitpodService().server.updateLoggedInUser(partialUser);
};

```

File: components/dashboard/src/data/organizations/org-members-query.ts
Comment (Siddhant-K-code) [inline components/dashboard/src/data/organizations/org-members-query.ts:30]: **Suggestion (non-blocking):** We can add more context to get more info about this check.

```suggestion
        /**
         * If no current org is set, disable query
         * This is to prevent making a request to the API when there is no organization selected.
         * This is the case when the user is first logging in, or when the user is on the settings page, and no organization is selected.
         */
        enabled: !!organization,
```

File: components/dashboard/src/onboarding/OnboardingStep.tsx
Comment (Siddhant-K-code) [inline components/dashboard/src/onboarding/OnboardingStep.tsx:17]: AFAIK these props are for Onboarding Forms.

`isLoading` looks ambiguous as we are already using in List Workspace ([code](https://github.com/gitpod-io/gitpod/blob/f2d868e7eaa111bf6308cb7c4acc085e7c3775e7/components/dashboard/src/workspaces/Workspaces.tsx#L30)). e.g. `isFormLoading`could be good. Or for quick :shipit: we can just add the comment for context.


```suggestion
type Props = {
  // The title of the form
  title: string;
  // The subtitle of the form
  subtitle: string;
  // Whether the form is valid
  isValid: boolean;
  // Whether the form is loading
  isLoading?: boolean;
  // A message to display if there is an error
  error?: string;
  // A function to call when the form is submitted
  onSubmit(): void;
};
```

File: components/dashboard/src/onboarding/StepOrgInfo.tsx
Comment (Siddhant-K-code) [inline components/dashboard/src/onboarding/StepOrgInfo.tsx:166]: Some Code cutting Suggestion 😉 
```suggestion
                            onChange={(e) => {
                                e.target.checked ? addSignupGoal(o.value) : removeSignupGoal(o.value);
                            }}
```

File: components/dashboard/src/onboarding/StepPersonalize.tsx
Comment (Siddhant-K-code) [inline components/dashboard/src/onboarding/StepPersonalize.tsx:11]: We are not using this anywhere in the file.

```suggestion
```

File: components/dashboard/src/onboarding/StepPersonalize.tsx
Comment (Siddhant-K-code) [inline components/dashboard/src/onboarding/StepPersonalize.tsx:32]: **nit-typo**

```suggestion
            subtitle="Customize your experience"
```

File: components/dashboard/src/onboarding/signup-goals.ts
Comment (Siddhant-K-code) [inline components/dashboard/src/onboarding/signup-goals.ts:22]: It looks like some other unicode character:

```suggestion
        { label: `Solve "works on my machine issue"`, value: "works_on_my_machine" },
```

File: components/dashboard/src/onboarding/job-roles.ts
Comment (Siddhant-K-code) [inline components/dashboard/src/onboarding/job-roles.ts:14]: **Suggestion(non-blocking):** 

Source: [w3schools](https://www.w3schools.com/whatis/whatis_fullstack.asp#:~:text=What%20is-,Full%20Stack,-%3F)

```suggestion
        { label: "Full Stack", value: "fullstack" },
```

**Comment by loujaybee** [comment]
Hey @mustard-mh + @filiptronicek, we are looking to get this pull request through quite soon. I know you might not know all the in's and out's of the codebase here, but could we help out @selfcontained with a quick review? 

If you can, do try and keep any comments actionable and clear in terms of what you are suggesting to change or update, with a preference to direct suggestions to swap, so that Brad can accept/reject more quickly ! 🎸 

... But I know that you folks know what you're doing, so I'll leave it with you! 🙏

File: components/dashboard/src/onboarding/StepUserInfo.tsx
Comment (filiptronicek) [inline components/dashboard/src/onboarding/StepUserInfo.tsx:84]: nit: I think some things in here are missing that would be nice to add:
- [`autoComplete`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete) - to make it easier for users to fill it in
- [`required`](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required) - make the inputs accessible by not just doing input validation, but signalling to the browser, that these input fields must be filled in

File: components/dashboard/src/onboarding/StepUserInfo.tsx
Comment (filiptronicek) [inline components/dashboard/src/onboarding/StepUserInfo.tsx:53]: nit: not everyone has a last name, we should take that into account [[1](https://deedpolloffice.com/change-name/law)] - examples: [someone I know](https://twitter.com/DasSurma), [a list](https://en.wikipedia.org/wiki/List_of_legally_mononymous_people).

Although there may never be someone without a last name to see this flow, what would they put here if they don't have it? 😄

**Comment by gtsiolis** [comment]
Looking at this now! 👀

File: components/dashboard/src/onboarding/StepOrgInfo.tsx
Comment (filiptronicek) [inline components/dashboard/src/onboarding/StepOrgInfo.tsx:104]: thought (ux): Can we get something like a <kbd>Skip</kbd> button here? Maybe also in the step before

**Comment by mustard-mh** [comment]
I tried with `?onboarding=force`, but didn't see this onboarding flow, did I miss any steps? @selfcontained 

https://bmh-onboarding-form.preview.gitpod-dev.com/?onboarding=force#https://github.com/gitpod-io/gitpod

File: components/dashboard/src/onboarding/StepOrgInfo.tsx
Comment (gtsiolis) [inline components/dashboard/src/onboarding/StepOrgInfo.tsx:155]: **suggestion**: We do have a [`Checkbox`](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/src/components/CheckBox.tsx) component that could be reused here.

**Review by gtsiolis — APPROVED** [review]
Thanks for making this happen, @selfcontained! 🔮 

💭 I'd leave comments with suggestions in individual lines, but for the sake of minimizing the distraction, noise, and [scope creep](https://en.wikipedia.org/wiki/Scope_creep) in this PR, I've added only one review comment about a component reusability.

❓ The only question that's possibly worth asking is whether we could skip the first + last name here and go with the proposal of the preferred name you mentioned in our recent discussion. Cc @loujaybee @phimae 

🛹  I've also posted a slightly improved version of the flow in a [relevant discussion](https://gitpod.slack.com/archives/C04PH56LZK5/p1677151360084819) (internal), that includes quite a few changes, including additional steps, better disabled states, and more. However, I'd not hold this back for the first iteration. Adding below three of the pages affected here in case we could grab any design bits that are easy to address here like copy changes, font styles, etc. See also [design thinking](https://gitpod.slack.com/archives/C04PH56LZK5/p1677151373474799?thread_ts=1677151360.084819&cid=C04PH56LZK5) (internal) behind these changes.

| Step 1 | Step 2 | Step 3 |
|-|-|-|
| Welcome | Welcome-2 | Welcome-3 |

Approving to unblock merging, holding in case you'd like to address any of the UX suggestions.

/hold

File: components/dashboard/src/components/ThemeSelector.tsx
Comment (filiptronicek) [reply components/dashboard/src/components/ThemeSelector.tsx:1]: ‼️ @selfcontained maybe something broke here? Now, visiting from the preview environment, I get a different layout in my preferences than in production (see the <kbd>Theme</kbd> section)
| gitpod.io | PE |
|----|----|
| image | image |

File: components/dashboard/src/onboarding/StepUserInfo.tsx
Comment (filiptronicek) [reply components/dashboard/src/onboarding/StepUserInfo.tsx:84]: Let's add this in a follow-up PR, as it's nothing blocking.

File: components/dashboard/src/onboarding/StepUserInfo.tsx
Comment (filiptronicek) [reply components/dashboard/src/onboarding/StepUserInfo.tsx:53]: Also, let's leave it to a follow-up.

**Comment by Siddhant-K-code** [comment]
> I tried with ?onboarding=force, but didn't see this onboarding flow, did I miss any steps? `@selfcontained`
>
> https://bmh-onboarding-form.preview.gitpod-dev.com/?onboarding=force#https://github.com/gitpod-io/gitpod

@mustard-mh, Related [Internal conversation](https://gitpod.slack.com/archives/C04PH56LZK5/p1677142301768459?thread_ts=1677116348.700349&cid=C04PH56LZK5)

**Comment by jakobhero** [comment]
thanks for working on this @selfcontained !🙏🏼 i'm having a look at the tracking of this now👀

File: components/dashboard/src/components/ThemeSelector.tsx
Comment (gtsiolis) [reply components/dashboard/src/components/ThemeSelector.tsx:1]: Great catch, @filiptronicek! ⚾

Comment (selfcontained) [reply components/dashboard/src/components/ThemeSelector.tsx:1]: Good catch, I did adjust things a bit to make it responsive for the onboarding flow, which is funky here. I'll fix it.

File: components/dashboard/src/data/current-user/update-mutation.ts
Comment (selfcontained) [reply components/dashboard/src/data/current-user/update-mutation.ts:19]: Yah, good thinking. I'm going to hold off on extracting these for now until there's another mutation that would use the function, but that would be a good way to share some logic.

File: components/dashboard/src/onboarding/OnboardingStep.tsx
Comment (selfcontained) [reply components/dashboard/src/onboarding/OnboardingStep.tsx:17]: Good point, the more I think about it, `isLoading` isn't very indicative of what's happening here, it's really indicating the form is saving, and puts it in a disabled state. I'll adjust it to something like `isSaving` - and probably need to add another check to avoid redundant submits if that flag is set.

File: components/dashboard/src/onboarding/StepOrgInfo.tsx
Comment (selfcontained) [reply components/dashboard/src/onboarding/StepOrgInfo.tsx:155]: I need to make one, this was the first checkbox I've added ui for, so wanted to build it out a little, and will extract it to a component in a followup. Checkboxes are always so different than other inputs, so I wasn't sure what our UX pattern might look like for them.

Comment (gtsiolis) [reply components/dashboard/src/onboarding/StepOrgInfo.tsx:155]: DEAL—Opened https://github.com/gitpod-io/gitpod/issues/16511 yesterday that mentions a checkbox component which is possibly worth splitting into a separate issues for each form element. 🏓

File: components/dashboard/src/onboarding/StepUserInfo.tsx
Comment (selfcontained) [reply components/dashboard/src/onboarding/StepUserInfo.tsx:53]: Yup, I'm on the same page as you. I anticipate this will need some changes. Fortunately we're not moving to store a distinct first/last name in the db, just a full name still.

Comment (gtsiolis) [reply components/dashboard/src/onboarding/StepUserInfo.tsx:53]: We could possibly skip the name split with the designs in https://github.com/gitpod-io/gitpod/pull/16501#pullrequestreview-1311386471. Cc @phimae

File: components/dashboard/src/onboarding/signup-goals.ts
Comment (mustard-mh) [inline components/dashboard/src/onboarding/signup-goals.ts:16]: Maybe make it (`CEDs`) possible to add a link here?

Comment (Siddhant-K-code) [reply components/dashboard/src/onboarding/signup-goals.ts:16]: @mustard-mh I think, He meant Cloud Dev Environments ([CDEs](https://www.gitpod.io/cde))

File: components/dashboard/src/onboarding/StepUserInfo.tsx
Comment (selfcontained) [reply components/dashboard/src/onboarding/StepUserInfo.tsx:84]: Added `required`, good call. `autoComplete` is `"on"` by default in browsers, so don't need to explicitly add specify it unless you want to disable it. 

I did add `type="submit` to the button, but most browsers don't require that for the autocomplete hueristic anymore (but good to be explicit there anyways). Some other conditions browsers use to decide if it should provide autocomplete that include things like, was the input empty initially (if not, it won't offer autocomplete), does the input have a `name` attribute (most browsers don't require this one anymore afaik, but it'd be a good thing for us to add in a followup)

File: components/dashboard/src/onboarding/signup-goals.ts
Comment (selfcontained) [reply components/dashboard/src/onboarding/signup-goals.ts:16]: A link would be great - going to leave that for a followup as I'd need to adjust a few things with how these labels render so it supports it.

Comment (mustard-mh) [reply components/dashboard/src/onboarding/signup-goals.ts:16]: > @mustard-mh I think, He meant Cloud Dev Environments ([CDEs](https://www.gitpod.io/cde))

Yep, I know it, but for users may don't know this shorthand new word 😂 just want to give them more context

File: components/dashboard/src/onboarding/StepOrgInfo.tsx
Comment (mustard-mh) [reply components/dashboard/src/onboarding/StepOrgInfo.tsx:104]: Agree, and also

- `Skip` to skip whole onboarding flow
- Another `Skip` to skip current step
- Maybe we can also provide a way to show how many steps we have, and where we are now
- Change the last step `Continue` button title to `Complete`?

**Pull request merged**
